integers and floats

Yunif3 opened the pull request:
**Num Type, Literal, and Operators have been split into [Int|Float] Type, [Int|Float] Literals, and [Int|Float] Arithmetic Operators.**

I/F will stand for [Int|Float] from now on.

**Type**
- Int type replaced Num type
- Float type was added
- Int operators will treat floats as wrong inputs now, and vice versa
- I/F type ascription can be made by inputting "I" or "F" respectively

**Literals**
- I/F Lit is represented as string now
- IntLit replaced NumLit
- FloatLit can be constructed by using the "." on IntLit
- Temporary backspace/delete cases were added to take care of solo-".", which will be taken care of by InvalidVar hole in the future
- Leading zeros and trailing zeros in I/F Lits are treated as valid inputs

**Arithmetic Operators**
- I/F have separate operators
- Float operators will have F in front of its name (ex. FPlus)
- Float arithmetic operators can be created by adding "." to Int arithmetic operator, and it can be reversed through backspace/delete
- I/F operators have been put into separate modules in DHExp.re

Commit:
first commit so that I don't have to restart again

Commit:
all Num types changed to Int types. Variable names involving Num are still kept at the moment

Commit:
introduced Float type HType

Commit:
added Float type to all the Type Files

Commit:
fixed some typos

Commit:
introduced FloatLit and Float arithmetic operators

Commit:
Float and FloatLit have been taken care of where Int and NumLit are present

Commit:
Compiles successfully with float and float type correctly shows up on the Hazel toplevel. I can't currently type float type yet though

Commit:
allowed the introduction of Float Ops by adding tokens to the parser and lexer

Commit:
allowed the construction of Float type ascriptions

Commit:
changed NumLit/numlit to IntLit/intlit

Commit:
Revert "allowed the construction of Float type ascriptions"

This reverts commit 177c138814eab3848aec6b4999ffea9428583881.

Commit:
Revert "changed NumLit/numlit to IntLit/intlit"

This reverts commit a7edbbc71e01b71ada6767418a6596195549e18f.

Commit:
sucessfully reverted to the right commit (float type ascription)

dm0n3y committed:
Merge remote-tracking branch 'origin/dev' into int_float

dm0n3y commented on src/hazelcore/Associator.re:
```diff
@@ -1,3 +1,4 @@
+// Anytime a new operator is added, the Skel____Lexer.mll and Skel____Parser.mll should be updated too
```
The parser file extension is `.mly`. Let's also refine this comment so that it's clear what the alternatives are, e.g., "the corresponding `Skel[Typ|Pat|Expr]Lexer.mll` and `Skel[Typ|Pat|Expr].mly` files should be updated too".

dm0n3y commented on src/hazelcore/semantics/dynamics/DHExp.re:
```diff
@@ -54,6 +63,7 @@ type t =
   | Ap(t, t)
   | BoolLit(bool)
   | NumLit(int)
+  | FloatLit(float)
```
Let's change `NumLit` to `IntLit` here and in `DHPat` too

dm0n3y commented on src/hazelcore/semantics/dynamics/Dynamics.re:
```diff
@@ -938,14 +958,33 @@ module Exp = {
+      }
+    | BinOp(NotInHole, FMinus as op, skel1, skel2)
+    | BinOp(NotInHole, FPlus as op, skel1, skel2)
+    | BinOp(NotInHole, FTimes as op, skel1, skel2) =>
```
You can condense these cases into a single pattern `BinOp(NotInHole, (FMinus | FPlus | FTimes) as op, skel1, skel2)`

dm0n3y commented on src/hazelcore/semantics/dynamics/DHExp.re:
```diff
@@ -5,15 +5,21 @@ type bin_num_op =
   | Times
+  | FMinus
+  | FPlus
+  | FTimes
```
Instead of adding float operators here, I'd rename `bin_num_op` to `bin_int_op` and add a separate `bin_float_op` type, then add a `BinFloatOp` case to `DHExp.t`

dm0n3y commented on src/hazelcore/semantics/dynamics/Dynamics.re:
```diff
@@ -1739,14 +1793,25 @@ module Evaluator = {
 
-  let eval_bin_num_op = (op: DHExp.bin_num_op, n1: int, n2: int): DHExp.t =>
+  let eval_bin_int_op =
+      (op: DHExp.bin_num_op, n1: int, n2: int): option(DHExp.t) =>
```
cf comment at top of `DHExp.re`

dm0n3y commented on src/hazelcore/semantics/dynamics/Dynamics.re:
```diff
@@ -1711,6 +1761,9 @@ module Evaluator = {
    */
+  /* TODO
+     need to include the case for float point numbers?
+      */
```
cf comment at top of `DHExp.re`

dm0n3y commented on src/hazelcore/Examples.re:
```diff
@@ -277,6 +277,35 @@ let qsort_example: UHExp.t = {
+
+     UHExp.[append_letline, EmptyLine, partition_letline, EmptyLine, qsort_line];
+   };
+   */
```
remove

dm0n3y replied:
bump

dm0n3y commented on src/hazelcore/semantics/UHExp.re:
```diff
@@ -58,6 +64,7 @@ and operand =
   | EmptyHole(MetaVar.t)
   | Var(ErrStatus.t, VarErrStatus.t, Var.t)
   | NumLit(ErrStatus.t, int)
+  | FloatLit(ErrStatus.t, float)
```
Let's also change the name of `NumLit` to `IntLit` here and in `UHPat.re`

dm0n3y commented on src/hazelcore/semantics/Action.re:
```diff
@@ -2186,6 +2232,9 @@ module Exp = {
       };
+    /* | Some(Dot) =>
+       let ze = ZExp.ZBlock.wrap()
+       Succeeded(SynDone((ze, HType.Float, u_gen))) */
```
remove

dm0n3y commented on src/hazelcore/layout/UHDoc.re:
```diff
@@ -182,6 +182,9 @@ let mk_Unit = (~steps: CursorPath.steps, ()): t =>
 let mk_Num = (~steps: CursorPath.steps, ()): t =>
   Delim.mk(~path=(steps, 0), "Num") |> annot_Operand(~sort=Typ);
```
Change references to `Num` in this module to use `Int` terminology

dm0n3y commented on src/hazelcore/semantics/Action.re:
```diff
@@ -2936,6 +2985,26 @@ module Exp = {
+        | UHExp.Plus => UHExp.FPlus
+        | UHExp.Minus => UHExp.FMinus
+        | UHExp.Times => UHExp.FTimes
+        | _ => oper
```
Return `Failed` on an operator that can't be converted

dm0n3y commented on src/hazelcore/semantics/Action.re:
```diff
@@ -2936,6 +2985,26 @@ module Exp = {
+    /* Making Float Operators from Int Operators */
+    /* This is for temporary testing. I should find better cursor */
+    /* TODO FIXME */
+    | (Construct(SChar(".")), ZOperator((pos, oper), seq)) =>
```
I would restrict this behavior to when the cursor is after the operator in question

dm0n3y replied:
You'll also need to add a similar case to `ana_perform_opseq`. Right now this editing behavior only works in synthetic position, where there's no expected type. You can see this by trying to convert an int operator into a float operator in the defining expression of an annotated let.

dm0n3y replied:
Let's also add the ability to convert a float operator back to an int operator using `Backspace`

dm0n3y commented on src/hazelcore/semantics/Action.re:
```diff
@@ -2936,6 +2985,26 @@ module Exp = {
+         switch (new_operator |> ZExp.place_after_operator) {
+         | Some(x) => x
+         | None => zop
+         }; */
```
Clean up comments

dm0n3y reviewed (changes requested):
Looks good overall! Requested some minor changes in comments.

There's also an issue where you can only type out floats if what you type matches OCaml's output for `string_of_float`. So you can't type `1.0` left-to-right in Hazel, nor can you can start with `1|0`, type dot, and get `1.0`. We should let the user type whatever they want. I think the full solution here would be to have `UHExp.FloatLit` take a pair of strings or ints (representing left and right sides of decimal point) and convert to float value as needed for evaluation (and maybe add some check in `Action.re` to ensure that the user-entered value can actually be parsed into an OCaml float). cc @cyrus-

Commit:
cleared up the case of turning int op to float op

Commit:
cleaned up comments and simplified cases

Commit:
made the comment more clear

Commit:
separated bin int ops with bin float ops

Commit:
fixed the case for int op to float op

Commit:
switched IntLit(int) => IntLit(string) and FloatLit(float) => FloatLit(float)

Commit:
allowed input of ints and floats with leading and trailing zeros.

Commit:
floats can now have dot as the leading or ending character

Commit:
float ".x" case correctly removes the whole thing and replaces it with hole

Commit:
Float op can be turned into Int Op using backspace

Commit:
Merge branch 'int_float' of github.com:hazelgrove/hazel into int_float

Commit:
changed the NumLit to IntLit after merging from git pull

Yunif3 commented:
I pulled from the remote branch and fixed all the conflicts.

- Added the ability to convert Float Op to Int Op using Backspace
- Added the ability to remove ".x" float numbers
- Allowed leading zeros in Int and Float by changing IntLit and FloatLit representation as strings

dm0n3y committed:
Merge remote-tracking branch 'origin/dev' into int_float

dm0n3y commented on src/hazelcore/semantics/HTyp.re:
```diff
@@ -3,21 +3,23 @@
-// Prod(Num, Num)  ==  (Num, Num)
-// (Prod(Num, Num), Bool)  ==  (Num, Num, Bool)
+// Prod(Int, Int)  ==  (Int, Int)
+// (Prod(Int, Int), Bool)  ==  (Int, Int, Bool)
```
This wasn't you but let's remove

dm0n3y commented on src/hazelcore/semantics/Statics.re:
```diff
@@ -828,10 +836,18 @@ module Exp = {
     | BinOp(NotInHole, Minus, skel1, skel2)
     | BinOp(NotInHole, Plus, skel1, skel2)
     | BinOp(NotInHole, Times, skel1, skel2) =>
```
nit: collapse these into `BinOp(NotInHole, Minus | Plus | Times, skel1, skel2)`

dm0n3y commented on src/hazelcore/semantics/Statics.re:
```diff
@@ -828,10 +836,18 @@ module Exp = {
+      }
+    | BinOp(NotInHole, FMinus, skel1, skel2)
+    | BinOp(NotInHole, FPlus, skel1, skel2)
+    | BinOp(NotInHole, FTimes, skel1, skel2) =>
```
nit: collapse

dm0n3y commented on src/hazelcore/semantics/ZExp.re:
```diff
@@ -80,7 +80,9 @@ let valid_cursors_operand: UHExp.operand => list(CursorPosition.t) =
-  | NumLit(_, n) => CursorPosition.text_cursors(IntUtil.num_digits(n))
+  | IntLit(_, n) => CursorPosition.text_cursors(IntUtil.num_digits_str(n))
+  | FloatLit(_, f) =>
+    CursorPosition.text_cursors(FloatUtil.num_digits_str(f))
```
Let's just use `String.length` here and remove these utility functions.

dm0n3y commented on src/hazelcore/semantics/dynamics/DHExp.re:
```diff
@@ -1,31 +1,58 @@
 
 [@deriving sexp]
-type bin_num_op =
+type bin_int_op =
```
Sorry to come back to this but let's package up the types `bin_int_op` and `bin_float_op` in modules, e.g., create modules `BinIntOp` and `BinFloatOp`, rename the types as `t`, move `of_int_op`/`of_float_op` inside respective modules and rename as `of_op`, etc. We started this division on @annie-anna's PR too for strings and string ops.

dm0n3y commented on src/hazelcore/semantics/dynamics/DHExp.re:
```diff
@@ -53,8 +87,10 @@ type t =
-  | NumLit(int)
-  | BinNumOp(bin_num_op, t, t)
+  | IntLit(string)
+  | FloatLit(string)
```
These constructor args should be `int` and `float`. The `DH*` types are the internal expressions used for evaluation. Don't need to maintain the user-edited string.

dm0n3y commented on src/hazelcore/semantics/dynamics/Dynamics.re:
```diff
@@ -1739,15 +1804,28 @@ module Evaluator = {
+    switch (op) {
+    | Minus => Some(IntLit(string_of_int(n1 - n2)))
+    | Plus => Some(IntLit(string_of_int(n1 + n2)))
+    | Times => Some(IntLit(string_of_int(n1 * n2)))
```
cf comment on constructor args in `DHExp` and `DHPat`. By making that change, you'll need to push the `int_of_string`/`float_of_string` conversions into `expand` functions rather than calling them here during evaluation. There, you should use `int_of_string_opt`/`int_of_float_opt` and return `DoesNotExpand` if the conversions fail.

dm0n3y commented on src/hazelcore/semantics/Action.re:
```diff
@@ -6,6 +6,9 @@ type operator_shape =
   | STimes
+  | SFMinus
+  | SFPlus
+  | SFTimes
```
Do we actually need these? Maybe we'd eventually want to directly construct via keyboard command, but even that seems like overkill to me. cc @cyrus-

dm0n3y commented on src/hazelcore/semantics/Action.re:
```diff
@@ -3154,19 +3242,29 @@ module Exp = {
+    /* For Float: (.x<|) ==> ( |_ )  */
+    | (Backspace, CursorE(OnText(2), FloatLit(_, f)))
+        when
+          FloatUtil.less_than_one_str(f) && FloatUtil.num_digits_str(f) == 2 =>
```
This needs to be a little more general. For example, right now we don't handle backspacing on `1|.` or deleting on `.|1`.

Yunif3 replied:
means delete the whole thing when there is only a dot left

dm0n3y commented on src/hazelcore/util/FloatUtil.re:
```diff
@@ -0,0 +1,36 @@
+  } else {
+    0;
+  };
+};
```
Let's move `leading_zeros` and `trailing_zeros` into `StringUtil` and simplify implementation to be `int`/`float`-agnostic, e.g., recursively increment some char index down/up the string until you encounter the first nonzero.

dm0n3y replied:
Also rename to `num_leading_zeros`/`num_trailing_zeros`

dm0n3y commented on src/hazelcore/util/IntUtil.re:
```diff
@@ -1 +1,9 @@
 let num_digits = (n: int): int => String.length(string_of_int(n));
+let num_digits_str = (n: string): int => String.length(n);
+let leading_zeros = (n: string): int => {
```
cf comment in `FloatUtil`

dm0n3y reviewed (changes requested):
Nice job, looks and feels good. A couple gaps w.r.t. deleting floats + some minor refactor requests.

Commit:
simplied cases, deleted comments.
replaced and removed num_digits_str with String.length

Commit:
better formatting

Commit:
added the case for "." in mk_syn/ana_text.
Backspace/Delete on ".x" with cursor anywhere produces a new hole

Commit:
removed unneccessary shapes

Commit:
made bin_int_op and bin_float_op into modules

Commit:
removed shapes for Float Ops.
Shape_of_operator now returns option type since float ops don't have a shape

Commit:
moved the zero helper functions to StringUtil.
Added edge cases to TextShape for Int and Float

Commit:
restored the dynamic files to have IntLit and FloatLit as int and float respectively.
Added the case for DoesNotExpand

Commit:
Merge branch 'int_float' of github.com:hazelgrove/hazel into int_float

Yunif3 commented:
Fixed the pointed brought up by the 2nd review.

- Removed unnecessary comments and simplified the code with better formatting/modularity
- Removed unnecessary Op shapes from Action.re
- Allowed the ".x" to become an empty hole when there is an action that leaves only "."
- moved the zero helper functions to StringUtil
- fixed the IntLit/FloatLit implementation in DH* files

dm0n3y commented on src/hazelcore/util/FloatUtil.re:
```diff
@@ -0,0 +1,8 @@
+  String.length(string_of_float(f));
+};
+
+let less_than_one_str = (f: string): bool => {
```
unused, remove

Yunif3 replied:
do you mean to remove both functions or just less_than_one_str?

dm0n3y replied:
whatever is unused

dm0n3y commented on src/hazelcore/semantics/dynamics/Dynamics.re:
```diff
@@ -1031,7 +1084,9 @@ module Exp = {
-    | NumLit(NotInHole, n) => Expands(NumLit(n), Num, delta)
+    | IntLit(NotInHole, n) => Expands(IntLit(int_of_string(n)), Int, delta)
+    | FloatLit(NotInHole, f) =>
+      Expands(FloatLit(float_of_string(f)), Float, delta)
```
Gotta do the `opt` version here too

dm0n3y commented on src/hazelcore/semantics/Action.re:
```diff
@@ -2152,12 +2183,18 @@ module Exp = {
         Succeeded(SynDone((ZExp.ZBlock.wrap(zhole), HTyp.Hole, u_gen)));
+      } else if (text == ".") {
+        let (zhole, u_gen) = u_gen |> ZExp.new_EmptyHole;
+        Succeeded(SynDone((ZExp.ZBlock.wrap(zhole), HTyp.Hole, u_gen)));
```
Add a TODO comment here indicating that this behavior should change once we have `InvalidVar` holes.

dm0n3y replied:
Also you need to add this case to `Pat.mk_[syn|ana]_text` as well

dm0n3y commented on src/hazelcore/semantics/Action.re:
```diff
@@ -2204,6 +2241,9 @@ module Exp = {
         let (zhole, u_gen) = u_gen |> ZExp.new_EmptyHole;
         Succeeded(AnaDone((ZExp.ZBlock.wrap(zhole), u_gen)));
+      } else if (text == ".") {
+        let (zhole, u_gen) = u_gen |> ZExp.new_EmptyHole;
```
Add TODO here as well (and in `Pat` module)

dm0n3y commented on src/hazelcore/semantics/Action.re:
```diff
@@ -3273,12 +3363,17 @@ module Exp = {
     | (Construct(SAsc), CursorE(_)) => Failed
 
+    /* It is a temporary fix so that a new hole isn't created when a dot is inputted by itself.
+       It will later be replaced by invalid var and handled by make syntax */
```
Temporary fix so that a new hole isn't created when a dot is inputted by itself (cf mk_[syn|ana]_text)
TODO remove once we have InvalidVar holes

dm0n3y replied:
Think you'll also need to add this case to `Pat.[syn|ana]_perform_operand`

Commit:
added cases and comments for Pat.mk_[syn|ana]_text

Commit:
fixing typos for wrong type

Commit:
added opt for [int|float]_of_string
